packages/asset-swapper/src/utils/market_operation_utils/balancer_utils.ts

```diff
@@ -42,10 +46,52 @@ export class BalancerPoolsCache {
         return Promise.race([this._getPoolsForPairAsync(takerToken, makerToken), timeout]);
     }
 
-    protected async _getPoolsForPairAsync(takerToken: string, makerToken: string): Promise<BalancerPool[]> {
+    public getCachedPoolAddressesForPair(
+        takerToken: string,
+        makerToken: string,
+        cacheExpiryMs?: number,
+    ): string[] | undefined {
+        const key = JSON.stringify([takerToken, makerToken]);
+        const value = this._cache[key];
+        if (cacheExpiryMs === undefined) {
+            return value === undefined ? [] : value.pools.map(pool => pool.id);
+        }
+        const minTimestamp = Date.now() - cacheExpiryMs;
+        if (value === undefined || value.timestamp < minTimestamp) {
+            return undefined;
+        } else {
+            return value.pools.map(pool => pool.id);
+        }
+    }
+
+    public howToSampleBalancer(
+        takerToken: string,
+        makerToken: string,
+        excludedSources: ERC20BridgeSource[],
+    ): { onChain: boolean; offChain: boolean } {
+        // If Balancer is excluded as a source, do not sample.
+        if (excludedSources.includes(ERC20BridgeSource.Balancer)) {
+            return { onChain: false, offChain: false };
+        }
+        const cachedBalancerPools = this.getCachedPoolAddressesForPair(takerToken, makerToken, ONE_DAY_MS);
+        // Sample Balancer on-chain (i.e. via the ERC20BridgeSampler contract) if:
+        // - Cached values are not stale
+        // - There is at least one Balancer pool for this pair
+        const onChain = cachedBalancerPools !== undefined && cachedBalancerPools.length > 0;
+        // Sample Balancer off-chain (i.e. via GraphQL query + `computeBalancerBuy/SellQuote`)
+        // if cached values are stale
+        const offChain = cachedBalancerPools === undefined;
+        return { onChain, offChain };
+    }
+
+    protected async _getPoolsForPairAsync(
+        takerToken: string,
+        makerToken: string,
+        cacheExpiryMs: number = FIVE_SECONDS_MS,
```

Comment:
Is this still intentionally low?

Reply:
yeah, in theory we should almost never be getting a cache hit for this function because we will only be calling it if `howToSampleBalancer` decides that the cache is too stale to use on-chain sampling. Leaving it in just in case something goes awry